### PR TITLE
fix($httpParamSerializerJQLike): Follow jQuery parameter serialization l...

### LINKED ISSUE #11551: $httpParamSerializerJQLike serializing differently than jQuery.param
There's a difference when serializing nested objects:

```js
jQuery.param({ myObject: { prop1: 5, prop2: { propA: 10 } } })
```
produces
```
myObject%5Bprop1%5D=5&myObject%5Bprop2%5D%5BpropA%5D=10
```

While `$httpParamSerializerJQLike` produces
```
myObject%5Bprop1%5D=5&myObject%5Bprop2%5D=%7B%22propA%22:10%7D
```

`ASP.NET Web API` understands `jQuery.param` but not `$httpParamSerializerJQLike`.

BTW, thanks for adding this feature, @pkozlowski-opensource! :smiley:

### PR DESCRIPTION
**Pull request by lgalfaso**
...ogic

Follow jQuery parameter serialization logic for nested objects.

Closes #11551

### CODE REVIEW

File: test/ng/httpSpec.js [hunk 1]
```diff
@@ -2014,6 +2014,12 @@ describe('$http param serializers', function() {
       expect(jqrSer({a: 'b', foo: {'bar': 'barv', 'baz': 'bazv'}})).toEqual('a=b&foo%5Bbar%5D=barv&foo%5Bbaz%5D=bazv');
                                                                            //a=b&foo[bar]=barv&foo[baz]=bazv
     });
+
+    it('should serialize nested objects by repeating param name with [kay] suffix', function() {
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: _kay_  -->  _KEY_

Comment (lgalfaso) [inline test/ng/httpSpec.js]: Done

File: src/ng/http.js [hunk 2]
```diff
@@ -34,7 +35,32 @@ function paramSerializerFactory(jQueryMode) {
...
 
     return parts.length > 0 ? parts.join('&') : '';
-  };
+  }
+
+  function jQueryLikeParamSerializer(params) {
+    if (!params) return '';
+    var parts = [];
+    serialize(params, '', true);
+    return parts.length > 0 ? parts.join('&') : '';
```
Comment (gkalpak) [inline src/ng/http.js]: Minor, but I think this can be just `return parts.join('&');`
In `paramSerializer()` too.

Comment (lgalfaso) [inline src/ng/http.js]: Done

**Comment by lgalfaso** [comment]
The travis failure looks like a flake

**Comment by gabrielmaldi** [comment]
I think we need to make a special case for date parameters:

```js
jQueryLikeParamSerializer({ dateFrom: new Date() }) === ""
```

File: src/ng/http.js [hunk 3]
```diff
@@ -18,23 +20,47 @@ function paramSerializerFactory(jQueryMode) {
...
+    serialize(params, '', true);
+    return parts.join('&');
+
+    function serialize(toSerialize, prefix, topLevel) {
+      if (toSerialize === null || isUndefined(toSerialize)) return;
+      if (isArray(toSerialize)) {
+        forEach(toSerialize, function(value) {
+          serialize(value, prefix + '[]');
+        });
+      } else if (isObject(toSerialize)) {
```
Comment (gabrielmaldi) [inline src/ng/http.js]: Maybe `} else if (isObject(toSerialize) && !isDate(toSerialize)) {`?

**Comment by lgalfaso** [comment]
@gabrielmaldi you are right, fixed this case

**Comment by gabrielmaldi** [comment]
I thought this would make it into `v1.4.0-rc.1` :cry:

File: src/ng/http.js [hunk 4]
```diff
@@ -11,6 +11,8 @@ var JSON_PROTECTION_PREFIX = /^\)\]\}',?\n/;
 
 function paramSerializerFactory(jQueryMode) {
 
+  return jQueryMode ? jQueryLikeParamSerializer : paramSerializer;
```
Comment (petebacondarwin) [inline src/ng/http.js]: If we are going to do this then we might as wel ditch the `paramSerializerFactory` altogether and just reference these two serializers directly in their respective providers.

**Comment by petebacondarwin** [comment]
One comment but otherwise LGTM

File: src/ng/http.js (hunk 4 above)
Comment (lgalfaso) [inline src/ng/http.js]: They only share `serializeValue`, maybe move a little more logic to this function but keep it as is?

Comment (gabrielmaldi) [inline src/ng/http.js]: It's already refactored in 2420a0a77e27b530dbb8c41319b2995eccf76791

**Pull request closed without merging**